# Trends of “Ask HN” and “Show HN” posts of Hacker News

This project compares two types of posts from [Hacker News](https://news.ycombinator.com/), a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either `Ask HN` or `Show HN`.

A user can ask the Hacker News community a specific question, such as “Ask HN: How to improve my website?” Likewise, users submit `Show HN` posts to show the Hacker News community a project, product, or just generally something interesting (e.g., “Show HN: Shanhu.io, a programming playground powered by e8vm”).

We'll compare these two types of posts to determine the following:
 * Do `Ask HN` or `Show HN` receive more comments on average?
 * Do posts created at a specific time receive more comments on average?

## Introduction

The data set used in this project was initially downloaded from Kaggle and was provided by Hacker News ([link])( https://www.kaggle.com/hacker-news/hacker-news-posts). After, has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampled from the remaining submissions.

In [1]:
# Import the data set from a csv file
from csv import reader

# Open the file and parse the data
opened_file = open('hacker_news.csv', encoding='utf-8')
read_file = reader(opened_file)

# Convert the read file into a list of lists and remove the first row of the data, which contains
# the column names
hn = list(read_file)
header = hn[0]
hn = hn[1:]

# Display the header and first five rows
print(header)
print()
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Description of the columns:
1. `id`: The unique identifier from Hacker News for the post
1. `title`: The title of the post
1. `url`: The URL that the posts links to, if it the post has a URL
1. `num_points`: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
1. `num_comments`: The number of comments that were made on the post
1. `author`: The username of the person who submitted the post
1. `created_at`: The date and time at which the post was submitted

## Extracting Ask HN and Show HN posts
First, we'll separate the posts into `Ask HN`, `Show HN` and others.

In [2]:
ask_posts = []
show_posts = []
other_posts = []

for post in hn:
    
    title = post[1]
    
    if title.lower().startswith('ask hn'):
        
        ask_posts.append(post)
        
    elif title.lower().startswith('show hn'):
        
        show_posts.append(post)
    else:
        other_posts.append(post)

print('Number of Ask HN posts: ' + str(len(ask_posts)))
print('Number of Show HN posts: ' + str(len(show_posts)))
print('Number of other HN posts: ' + str(len(other_posts)))
# print(*ask_posts[:10], '\n', sep='\n')
# print(*show_posts[:10], '\n', sep='\n')
# print(*other_posts[:10], '\n', sep='\n')

Number of Ask HN posts: 1744
Number of Show HN posts: 1162
Number of other HN posts: 17194


## Calculating the average number of comments for Ask HN and Show HN posts
Now let's determine which type of posts receives more comments on average.

In [3]:
avg_ask_comments = sum([int(post[4]) for post in ask_posts]) / len(ask_posts)
avg_show_comments = sum([int(post[4]) for post in show_posts]) / len(show_posts)

print('Average number of comments:')
print(' - Ask HN posts: {:.1f}'.format(avg_ask_comments))
print(' - Show HN posts: {:.1f}'.format(avg_show_comments))

Average number of comments:
 - Ask HN posts: 14.0
 - Show HN posts: 10.3


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive about 10. Since ask posts are more likely to receive comments (36% more), we'll focus our remaining analysis just on these posts.

## Finding the amount of Ask HN posts and comments by hour created
Next, we'll determine if we can maximize the number of comments an ask post receives by creating it at a specific time. First, we'll find the number of ask posts made during each hour of the day, along with the number of comments those posts received.

In [4]:
import datetime as dt

counts_by_hour = {}
comments_by_hour = {}

for post in ask_posts:

    # Convert date and time strings to datetime objects and get the hour
    hour = dt.datetime.strptime(post[6], '%m/%d/%Y %H:%M').hour

    # Calculate the amount of posts created in each hour of the day, along with the number of comments received
    if hour in counts_by_hour:
        
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += int(post[4])
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = int(post[4])

print('Amount of posts created in each hour of the day:')
print(*sorted(counts_by_hour.items()), sep='\n')
print()
print('Number of comments received in each hour of the day:')
print(*sorted(comments_by_hour.items()), sep='\n')
print()
print('Maximum number of posts in an hour: ' + str(max(counts_by_hour.values())))
print('Maximum number of comments in an hour: ' + str(max(comments_by_hour.values())))

Amount of posts created in each hour of the day:
(0, 55)
(1, 60)
(2, 58)
(3, 54)
(4, 47)
(5, 46)
(6, 44)
(7, 34)
(8, 48)
(9, 45)
(10, 59)
(11, 58)
(12, 73)
(13, 85)
(14, 107)
(15, 116)
(16, 108)
(17, 100)
(18, 109)
(19, 110)
(20, 80)
(21, 109)
(22, 71)
(23, 68)

Number of comments received in each hour of the day:
(0, 447)
(1, 683)
(2, 1381)
(3, 421)
(4, 337)
(5, 464)
(6, 397)
(7, 267)
(8, 492)
(9, 251)
(10, 793)
(11, 641)
(12, 687)
(13, 1253)
(14, 1416)
(15, 4477)
(16, 1814)
(17, 1146)
(18, 1439)
(19, 1188)
(20, 1722)
(21, 1745)
(22, 479)
(23, 543)

Maximum number of posts in an hour: 116
Maximum number of comments in an hour: 4477


The period when most posts were created was between 15:00 and 16:00 (116 posts). The posts made on this period also received the most number of comments (4477).

## Calculating the average number of comments for Ask HN posts by hour
Now we'll calculate the average number of comments ask posts created at each hour of the day receive.

In [5]:
avg_by_hour = [(hour, comments_by_hour[hour] / counts_by_hour[hour]) for hour in counts_by_hour]
avg_by_hour

[(9, 5.5777777777777775),
 (13, 14.741176470588234),
 (10, 13.440677966101696),
 (14, 13.233644859813085),
 (16, 16.796296296296298),
 (23, 7.985294117647059),
 (12, 9.41095890410959),
 (17, 11.46),
 (15, 38.5948275862069),
 (21, 16.009174311926607),
 (20, 21.525),
 (2, 23.810344827586206),
 (18, 13.20183486238532),
 (3, 7.796296296296297),
 (5, 10.08695652173913),
 (19, 10.8),
 (1, 11.383333333333333),
 (22, 6.746478873239437),
 (8, 10.25),
 (4, 7.170212765957447),
 (0, 8.127272727272727),
 (6, 9.022727272727273),
 (7, 7.852941176470588),
 (11, 11.051724137931034)]

## Sorting and printing values from a list of lists

In [6]:
avg_by_hour = sorted(avg_by_hour, key = lambda x: x[1], reverse=True)

print('Top 5 average number of comments received per post:')

for hour, avg in avg_by_hour[:5]:

    print('{:02}:00: {:.1f}'.format(hour, avg))

Top 5 average number of comments received per post:
15:00: 38.6
02:00: 23.8
20:00: 21.5
16:00: 16.8
21:00: 16.0


Again, the period with the highest average number of comments received per post was between 15:00 and 16:00 (38.6 comments/post).

## Conclusions
This project compared data from ask and show posts from Hacker News to determine which type of post and time received the most comments on average. We concluded that the ask posts created between 15:00 and 16:00 had the maximum average number of comments. However, it should be noted that the data set we analyzed excluded posts without any comments.